diff --git a/include/triton/Dialect/Triton/IR/TritonOps.td b/include/triton/Dialect/Triton/IR/TritonOps.td
index e9f892da0..07b83751f 100644
--- a/include/triton/Dialect/Triton/IR/TritonOps.td
+++ b/include/triton/Dialect/Triton/IR/TritonOps.td
@@ -1410,5 +1410,30 @@ def TT_DescriptorScatterOp : TT_Op<"descriptor_scatter", [TT_DescriptorStoreLike
   let hasVerifier = 1;
 }
 
+def TT_AnnotationOp : TT_Op<"annotation", [Pure, MemoryEffects<[MemWrite]>]> {
+  let summary = "Annotate a tensor with key-value attribute pairs";
+  let description = [{
+    `tt.annotation` operation can be used to annotate a tensor with
+    key-value attribute pairs.
+
+    Example:
+    ```mlir
+    tt.annotation %target {key : val}
+    ```
+  }];
+  let arguments = (ins TT_Tensor:$src);
+  let assemblyFormat = [{
+    $src attr-dict `:` type($src)
+  }];
+}
+
+def TT_CustomSyncOp : TT_Op<"custom_sync",  [Pure, MemoryEffects<[MemWrite]>]> {
+  let summary = "self-defined custom sync operation";
+  let description = [{
+    `tt.custom_sync` triton custom sync op is designed to pass self-defined custom sync operation.
+  }];
+  let arguments = (ins StrAttr:$op_name, StrAttr:$mode_or_sender, I32Attr:$id);
+  let assemblyFormat = "$op_name attr-dict";
+}
 
 #endif // Triton_OPS
