diff --git a/python/triton/_utils.py b/python/triton/_utils.py
index d7d2455f0..e61c8a17a 100644
--- a/python/triton/_utils.py
+++ b/python/triton/_utils.py
@@ -50,8 +50,8 @@ def validate_block_shape(shape: List[int]):
     for i, d in enumerate(shape):
         if not isinstance(d, int):
             raise TypeError(f"Shape element {i} must have type `constexpr[int]`, got `constexpr[{type(d)}]")
-        if not is_power_of_two(d):
-            raise ValueError(f"Shape element {i} must be a power of 2")
+        # if not is_power_of_two(d):
+        #     raise ValueError(f"Shape element {i} must be a power of 2")
         numel *= d
 
     if numel > TRITON_MAX_TENSOR_NUMEL:
