diff --git a/python/src/ir.h b/python/src/ir.h
index b27c41a73..d1822002f 100644
--- a/python/src/ir.h
+++ b/python/src/ir.h
@@ -2,6 +2,14 @@
 #include "mlir/IR/Builders.h"
 #include "triton/Tools/Sys/GetEnv.hpp"
 #include <memory>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/stl_bind.h>
+
+namespace py = pybind11;
+
+using namespace mlir;
+using namespace triton;
 
 // A custom op builder that keeps track of the last location
 class TritonOpBuilder {
@@ -88,3 +96,7 @@ private:
   bool lineInfoEnabled =
       !mlir::triton::tools::getBoolEnv("TRITON_DISABLE_LINE_INFO");
 };
+
+namespace ir {
+extern py::class_<TritonOpBuilder> *getBuilderClass();
+} // namespace ir
\ No newline at end of file
