diff --git a/python/src/ir.cc b/python/src/ir.cc
index d79a9e70f..2ea7c4db3 100644
--- a/python/src/ir.cc
+++ b/python/src/ir.cc
@@ -245,6 +245,13 @@ py::list getTensorDescMetadata(ModuleOp &mod) {
 
 } // anonymous namespace
 
+namespace ir {
+
+static py::class_<TritonOpBuilder> *builderClassPtr = nullptr;
+py::class_<TritonOpBuilder> *getBuilderClass() { return builderClassPtr; }
+
+} // namespace ir
+
 /*****************************************************************************/
 /* Python bindings for ir                                                    */
 /*****************************************************************************/
@@ -390,6 +397,10 @@ void init_triton_ir(py::module &&m) {
       .def("param_types", [](FunctionType &self) {
         return std::vector<Type>(self.getInputs().begin(),
                                  self.getInputs().end());
+           })
+      .def("result_types", [](FunctionType &self) {
+        return std::vector<Type>(self.getResults().begin(),
+                                 self.getResults().end());
       });
 
   py::class_<Location>(m, "location", py::module_local())
@@ -515,6 +526,7 @@ void init_triton_ir(py::module &&m) {
   py::class_<IntegerAttr, Attribute>(m, "integer_attr", py::module_local());
   py::class_<BoolAttr, Attribute>(m, "bool_attr", py::module_local());
   py::class_<UnitAttr, Attribute>(m, "unit_attr", py::module_local());
+  py::class_<StringAttr, Attribute>(m, "string_attr", py::module_local());
 
   // Ops
   py::class_<OpState>(m, "OpState", py::module_local())
@@ -777,8 +789,10 @@ void init_triton_ir(py::module &&m) {
 
   py::class_<OpBuilder::InsertPoint>(m, "InsertPoint", py::module_local());
 
-  py::class_<TritonOpBuilder>(m, "builder", py::module_local(),
-                              py::dynamic_attr())
+  static py::class_<TritonOpBuilder> builderClass(
+      m, "builder", py::module_local(), py::dynamic_attr());
+  ir::builderClassPtr = &builderClass;
+  builderClass
       .def(py::init<MLIRContext *>())
       .def("get_op_builder", &TritonOpBuilder::getBuilder, ret::reference)
       // getters
