diff --git a/setup.py b/setup.py
index dffc45ab1..a4fe0b8a8 100644
--- a/setup.py
+++ b/setup.py
@@ -182,6 +182,13 @@ class Package:
 # json
 def get_json_package_info():
     url = "https://github.com/nlohmann/json/releases/download/v3.11.3/include.zip"
+    local_json_path = os.environ.get("JSON_PATH34", "")
+    if local_json_path != "":
+        if os.path.exists(local_json_path):
+            url = "file://" + local_json_path
+            print(f"JSON_PATH34 {local_json_path} exists. url={url}")
+        else:
+            print(f"JSON_PATH34 {local_json_path} does not exist.")
     return Package("json", "", url, "JSON_INCLUDE_DIR", "", "JSON_SYSPATH")
 
 
@@ -244,6 +251,13 @@ def get_llvm_package_info():
     # Create a stable symlink that doesn't include revision
     sym_name = f"llvm-{system_suffix}"
     url = f"https://oaitriton.blob.core.windows.net/public/llvm-builds/{name}.tar.gz"
+    local_llvm_tgz_path = os.environ.get("LLVM_TGZ_PATH34", "")
+    if local_llvm_tgz_path != "":
+        if os.path.exists(local_llvm_tgz_path):
+            url = "file://" + local_llvm_tgz_path
+            print(f"LLVM_TGZ_PATH34 {local_llvm_tgz_path} exists. url={url}")
+        else:
+            print(f"LLVM_TGZ_PATH34 {local_llvm_tgz_path} does not exist.")
     return Package("llvm", name, url, "LLVM_INCLUDE_DIRS", "LLVM_LIBRARY_DIR", "LLVM_SYSPATH", sym_name=sym_name)
 
 
@@ -528,6 +542,37 @@ class CMakeBuild(build_ext):
         subprocess.check_call(["cmake", "--build", "."] + build_args, cwd=cmake_dir)
         subprocess.check_call(["cmake", "--build", ".", "--target", "mlir-doc"], cwd=cmake_dir)
 
+        if(check_env_flag("IS_NOT_PUBLISH","1")):
+            cmake_build_dir = get_cmake_dir()
+            tools_src_dir = os.path.join(
+                cmake_build_dir,
+                "third_party/dicp_triton/third_party/triton_shared/tools/triton-shared-opt/"
+            )
+            dicp_opt_src = os.path.join(
+                cmake_build_dir,
+                "third_party/dicp_triton/tools/dicp_triton_opt/"
+            )
+            shutil.copy(
+                os.path.join(tools_src_dir, "triton-shared-opt"),
+                os.path.join(extdir, "triton-shared-opt-v3_4")
+            )
+            shutil.copy(
+                os.path.join(dicp_opt_src, "dicp_opt"),
+                extdir
+            )
+
+        if 'bdist_wheel' in sys.argv:
+            bishengir_path=os.path.join(extdir, "bishengir")
+            os.makedirs(bishengir_path, exist_ok=True)
+            bisheng_install_path = os.environ.get("BISHENG_INSTALL_PATH", "")
+            if bisheng_install_path=="":
+                print(f"bisheng_install_path is empty")
+            else:
+                print(f"bisheng_install_path {bisheng_install_path}")
+                if os.path.isdir(bisheng_install_path):
+                    shutil.copytree(bisheng_install_path, bishengir_path, dirs_exist_ok=True)
+                else:
+                    raise Exception("bisheng_install_path is not a directory")
 
 def download_and_copy_dependencies():
     nvidia_version_path = os.path.join(get_base_dir(), "cmake", "nvidia-toolchain-version.json")
@@ -649,6 +694,9 @@ def get_packages():
             # Install the contents of each backend's `tools` directory into
             # `triton.tools.extra`.
             for x in os.listdir(backend.tools_dir):
+                if x == "CMakeLists.txt":
+                    print(f"get_packages: skipping package for CMakeLists.txt: {x}", file=sys.stderr)
+                    continue
                 yield f"triton.tools.extra.{x}"
 
     if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
@@ -795,15 +843,15 @@ PYTHON_CLASSIFIERS = [
 CLASSIFIERS = BASE_CLASSIFIERS + PYTHON_CLASSIFIERS
 
 setup(
-    name=os.environ.get("TRITON_WHEEL_NAME", "triton"),
+    name=os.environ.get("TRITON_WHEEL_NAME", "dlcompiler"),
     version=TRITON_VERSION,
-    author="Philippe Tillet",
-    author_email="phil@openai.com",
+    author="Deeplink",
     description="A language and compiler for custom Deep Learning operations",
     long_description="",
     install_requires=[
         "setuptools>=40.8.0",
         "importlib-metadata; python_version < '3.10'",
+        "torch_npu>=2.6.0",
     ],
     packages=list(get_packages()),
     package_dir=dict(get_package_dirs()),
@@ -824,7 +872,7 @@ setup(
     zip_safe=False,
     # for PyPI
     keywords=["Compiler", "Deep Learning"],
-    url="https://github.com/triton-lang/triton/",
+    url="https://github.com/DeepLink-org/DLCompiler/",
     python_requires=PYTHON_REQUIRES,
     classifiers=CLASSIFIERS,
     test_suite="tests",
